core/drivers/hi16xx_uart.c

```diff
@@ -76,16 +79,69 @@
 #define UART_USR_RFNE_BIT	3	/* Receive FIFO not empty bit */
 #define UART_USR_RFF_BIT	4	/* Receive FIFO full bit */
 
-void hi16xx_uart_flush(vaddr_t base)
+static void hi16xx_uart_flush(struct serial_chip *chip)
 {
+	struct hi16xx_uart_data *pd = container_of(chip,
```

Comment:
All these functions would be a bit less complicated with a helper function like:
```
static vaddr_t chip_to_base(struct serial_chip *chip)
{
        struct hi16xx_uart_data *pd =
                container_of(chip, struct hi16xx_uart_data, chip);

        return io_pa_or_va(&pd->base);
}
```